File: tzone-accounting/frontend/src/core/hooks.ts
/**
 * The extension bus — the client half of the same idea as the backend's `HookBus`.
 *
 * Modules extend one another only through named hooks. `invoicing` contributes a posting rule;
 * `documents` collects posting rules. Neither imports the other, so adding the tenth document
 * type changes nothing that already exists.
 */

interface Handler {
  sequence: number;
  order: number;
  module: string;
  fn: (...args: never[]) => unknown;
}

export class HookBus {
  private handlers = new Map<string, Handler[]>();
  private counter = 0;
  private currentModule = "core";

  bindingModule(key: string): void {
    this.currentModule = key;
  }

  on(hook: string, fn: (...args: never[]) => unknown, sequence = 10): void {
    const list = this.handlers.get(hook) ?? [];
    list.push({ sequence, order: ++this.counter, module: this.currentModule, fn });
    list.sort((a, b) => a.sequence - b.sequence || a.order - b.order);
    this.handlers.set(hook, list);
  }

  /** Gather one contribution per handler, dropping `undefined`/`null`. */
  collect<T>(hook: string, ...args: unknown[]): T[] {
    const results: T[] = [];
    for (const handler of this.handlers.get(hook) ?? []) {
      const value = (handler.fn as (...a: unknown[]) => T | undefined)(...args);
      if (value !== undefined && value !== null) results.push(value);
    }
    return results;
  }

  /** Fire and forget. */
  emit(hook: string, ...args: unknown[]): void {
    for (const handler of this.handlers.get(hook) ?? []) {
      (handler.fn as (...a: unknown[]) => unknown)(...args);
    }
  }

  async emitAsync(hook: string, ...args: unknown[]): Promise<void> {
    for (const handler of this.handlers.get(hook) ?? []) {
      await (handler.fn as (...a: unknown[]) => unknown)(...args);
    }
  }

  describe(): Record<string, string[]> {
    const out: Record<string, string[]> = {};
    for (const [hook, list] of [...this.handlers.entries()].sort()) {
      out[hook] = list.map((h) => `${h.module} (seq ${h.sequence})`);
    }
    return out;
  }
}

File: tzone-accounting/backend/app/core/hooks.py
"""The extension bus.

Modules extend each other only through named hooks — never by importing one another's
internals. A module that wants to react to something calls `hooks.on(...)`; a module that wants
to be extended calls `hooks.emit(...)` or `hooks.collect(...)`. Adding a listener never requires
editing the emitter, which is what lets a hundred modules coexist.

Handlers run in ascending `sequence`, so a later module can deliberately run before or after an
earlier one without either knowing about the other.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass(order=True)
class _Handler:
    sequence: int
    order: int
    module: str = field(compare=False)
    fn: Callable[..., Any] = field(compare=False)


class HookBus:
    def __init__(self) -> None:
        self._handlers: dict[str, list[_Handler]] = {}
        self._counter = 0
        self._current_module = "core"

    def binding_module(self, module_key: str) -> None:
        """Set by the registry while a module is being loaded, for diagnostics."""
        self._current_module = module_key

    def on(self, name: str, fn: Callable[..., Any], sequence: int = 10) -> None:
        self._counter += 1
        self._handlers.setdefault(name, []).append(
            _Handler(sequence=sequence, order=self._counter, module=self._current_module, fn=fn)
        )
        self._handlers[name].sort()

    def emit(self, name: str, **kwargs: Any) -> None:
        """Fire-and-forget notification. Return values are ignored."""
        for handler in self._handlers.get(name, []):
            handler.fn(**kwargs)

    def collect(self, name: str, **kwargs: Any) -> list[Any]:
        """Gather one contribution per handler, dropping `None`."""
        results = []
        for handler in self._handlers.get(name, []):
            value = handler.fn(**kwargs)
            if value is not None:
                results.append(value)
        return results

    def chain(self, name: str, value: Any, **kwargs: Any) -> Any:
        """Pass a value through every handler, each returning the next value.

        Used where modules refine a shared result — e.g. enriching a pushed record before it
        is written.
        """
        for handler in self._handlers.get(name, []):
            value = handler.fn(value, **kwargs)
        return value

    def describe(self) -> dict[str, list[str]]:
        return {
            name: [f"{h.module} (seq {h.sequence})" for h in handlers]
            for name, handlers in sorted(self._handlers.items())
        }
